File: leetcode/data_structures/array/53.py
from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:

        mx = nums[0]
        mn = min(0, nums[0])

        for i in range(1, len(nums)):
            nums[i] += nums[i-1]
            mx = max(mx, nums[i] - mn)
            mn = min(mn, nums[i])
        
        return mx

File: leetcode/dynamic_programming/kadane_algorithm/53.go
package main

func Max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func Min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxSubArray(nums []int) int {
	var n int = len(nums)
	var acc, max, min int = nums[0], nums[0], Min(0, nums[0])

	for i := 1; i < n; i++ {
		acc += nums[i]
		max = Max(max, acc-min)
		min = Min(min, acc)
	}

	return max
}
